Plutarch/Enum.hs

```diff
@@ -0,0 +1,120 @@
+module Plutarch.Enum (
+  -- * Type classes
+  PCountable (..),
+  PEnumerable (..),
+) where
+
+import Data.Kind (Type)
+import Plutarch.Bool (POrd, pif, (#==))
+import Plutarch.Integer (PInteger)
+import Plutarch.Internal (
+  S,
+  Term,
+  phoistAcyclic,
+  (#),
+  (#$),
+  (:-->),
+ )
+import Plutarch.Internal.Other (pfix, pto)
+import Plutarch.Internal.PLam (plam)
+import Plutarch.Positive (PPositive)
+
+{- | A notion of \'next\' value. More formally, instances of this type class are
+discrete linear orders with no maximal element.
+
+= Laws
+
+1. @x /= psuccessor x@
+2. @y < x@ @=@ @psuccessor y <= x@
+3. @x < psuccessor y@ @=@ @x <= y@
+
+If you define 'psuccessorN', you must also ensure the following hold; the
+default implementation ensures this.
+
+4. @psuccessorN 1@ @=@ @psuccessor@
+5. @psuccessorN n . psuccessorN m@ @=@ @psuccessorN (n + m)@
+
+Law 1 ensures no value is its own successor. Laws 2 and 3 ensure that there
+are no \'gaps\': every value is \'reachable\' from any lower value by a
+finite number of applications of 'successor'.
+
+@since WIP
+-}
+class POrd a => PCountable (a :: S -> Type) where
+  -- | @since WIP
+  psuccessor :: forall (s :: S). Term s (a :--> a)
+
+  -- | @since WIP
+  {-# INLINEABLE psuccessorN #-}
+  psuccessorN :: forall (s :: S). Term s (PPositive :--> a :--> a)
+  psuccessorN = phoistAcyclic $ plam $ \n x -> go n # (psuccessor # x) # 1
+    where
+      go ::
+        forall (s' :: S).
+        Term s' PPositive ->
+        Term s' (a :--> PPositive :--> a)
+      go limit = pfix #$ plam $ \self acc count ->
```

Comment:
This is getting recursed on script. Is this more favorable to having it generate script that just runs `psuccessor` n times? 

Latter will make script size grow linearly on the argument `n` but with smaller execution cost.

Reply:
Maybe it's better to just not provide default instance here as many types like `PInteger` `PPositive` have much better way of dealing with `psuccessorN` than what's provided as default

Reply:
Running `psuccessor` N times would require N to be a constant, which is a bit limiting. 

I reckon that a default instance, assuming we note that it's not particularly efficient, isn't a problem. All of our existing instances are hand-rolled anyway.